### PR TITLE
Test Failures | Validate numpy.dtype for read/write arrays in grpc interpreter

### PR DESCRIPTION
**Pull request by MounikaBattu17**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nidaqmx-python/blob/master/CONTRIBUTING.md).

### What does this Pull Request accomplish?
This PR allows grpc_interpreter to validate `dtype` of a numpy array. 
Following the [Bug 2390189](https://dev.azure.com/ni/DevCentral/_workitems/edit/2390189): GrpcStubInterpreter reads/writes invalid data when NumPy array has wrong dtype context here, 
Unlike library interpreter, grpc interpreter doesn't raise error when wrong dtype is passed to the numpy array. Thus

- Created a custom method `_validate_array_dtype` in `_grpc_interpreter.py.mako` to validate expected and actual `dtype` of a `numpy` array of read/write functions.
- Create a custom `Error` to be raised when actual dtype doesn't match with expected dtype.
- Added helpers method `get_numpy_array_params` in `interpreter_helpers.py` to get `read/write` array parameters.
- Since `numpy.generic` can take in any numpy dtype, so excluded it from validation.

### Why should this Pull Request be merged?
This PR fixes the [Bug 2390189](https://dev.azure.com/ni/DevCentral/_workitems/edit/2390189): GrpcStubInterpreter reads/writes invalid data when NumPy array has wrong dtype

### What testing has been done?
- Ran `poetry run pytest` -> Existing test results maintained
- Passed in incorrect dtypes to the code snippet in the first comment given below and observed the Custom Error been raised by the grpc interpreter.

### CODE REVIEW

**Comment by MounikaBattu17** [comment]
Ran this code snippet by changing the items given in comments.

```
import nidaqmx
import numpy
from nidaqmx.constants import LineGrouping
from nidaqmx.stream_writers import DigitalSingleChannelWriter

with nidaqmx.Task() as task:
    task.do_channels.add_do_chan(
        "Dev1/port1/line0:3", line_grouping=LineGrouping.CHAN_FOR_ALL_LINES
    )

    writer = DigitalSingleChannelWriter(task.out_stream, auto_start=True)
    write_array = numpy.array([1, 2, 4, 8], dtype=numpy.uint8)  # passed incorrect dtype like numpy.uint32 etc
    print(writer.write_many_sample_port_byte(write_array)) # changed write_many_sample_port_byte to other write functions
```

**Review by zhindes — APPROVED** [review]

File: src/codegen/templates/_grpc_interpreter.py.mako [hunk 1]
```diff
@@ -159,3 +165,8 @@ def _assign_numpy_array(numpy_array, grpc_array):
         numpy_array.flat[:grpc_array_size] = numpy.frombuffer(grpc_array, dtype=numpy.uint8)
     else:
         numpy_array.flat[:grpc_array_size] = grpc_array
+
+def _validate_array_dtype(numpy_array, expected_numpy_array_dtype):
+    """Raises Typerror if array type doesn't match with expected numpy.dtype"""
+    if expected_numpy_array_dtype != numpy.generic and numpy_array.dtype != expected_numpy_array_dtype:
+        raise errors.DaqError(f"TypeError: array must have data type {expected_numpy_array_dtype}", DAQmxErrors.UNKNOWN)
```
Comment (bkeryan) [inline src/codegen/templates/_grpc_interpreter.py.mako:172]: This raises `DaqError`, not `TypeError`.
https://docs.python.org/3/library/exceptions.html#TypeError

```suggestion
    """Raises TypeError if array type doesn't match with expected numpy.dtype"""
    if expected_numpy_array_dtype != numpy.generic and numpy_array.dtype != expected_numpy_array_dtype:
        raise TypeError(f"array must have data type {expected_numpy_array_dtype}")
```

File: src/codegen/templates/_grpc_interpreter.py.mako [hunk 2]
```diff
@@ -17,6 +17,7 @@ from nidaqmx._base_interpreter import BaseInterpreter
 from nidaqmx._stubs import nidaqmx_pb2 as grpc_types
 from nidaqmx._stubs import nidaqmx_pb2_grpc as nidaqmx_grpc
 from nidaqmx._stubs import session_pb2 as session_grpc_types
+from nidaqmx.error_codes import DAQmxErrors
```
Comment (bkeryan) [inline src/codegen/templates/_grpc_interpreter.py.mako:20]: You don't need this to raise `TypeError`
```suggestion
```

**Review by bkeryan — APPROVED** [review]
Approved with suggestions

File: src/codegen/templates/_grpc_interpreter.py.mako (hunk 1 above)
Comment (MounikaBattu17) [reply src/codegen/templates/_grpc_interpreter.py.mako]: Removed DaqError & placed TypeError

**Pull request merged**